src/DotVVM.Framework/Compilation/Javascript/Ast/JsAstHelpers.cs

```diff
@@ -79,17 +79,37 @@ public static IEnumerable<JsExpression> GetLeafResultNodes(this JsExpression exp
             return node;
         }
 
-        /// Wraps the expression in `dotvvm.evaluator.wrapExpression` if needed
-        public static JsExpression EnsureObservableWrapped(this JsExpression expression) =>
+        /// Wraps the expression in `dotvvm.evaluator.wrapKnockoutExpression` if needed
+        public static JsExpression EnsureObservableWrapped(this JsExpression expression)
+        {
             // It's not needed to wrap if none of the descendants return an observable
-            !expression.DescendantNodes().Any(n => (n.HasAnnotation<ResultIsObservableAnnotation>() && !n.HasAnnotation<ShouldBeObservableAnnotation>()) || n.HasAnnotation<ObservableUnwrapInvocationAnnotation>()) ?
-
-                expression.WithAnnotation(ShouldBeObservableAnnotation.Instance) :
+            if (!expression.DescendantNodes().Any(n => (n.HasAnnotation<ResultIsObservableAnnotation>() && !n.HasAnnotation<ShouldBeObservableAnnotation>()) || n.HasAnnotation<ObservableUnwrapInvocationAnnotation>()))
+            {
+                return expression.WithAnnotation(ShouldBeObservableAnnotation.Instance);
+            }
+            else
+            {
+                var arguments = new List<JsExpression>(3);
 
-                new JsIdentifierExpression("dotvvm").Member("evaluator").Member("wrapKnockoutExpression").Invoke(new JsFunctionExpression(
+                arguments.Add(new JsFunctionExpression(
                     parameters: Enumerable.Empty<JsIdentifier>(),
-                    bodyBlock: new JsBlockStatement(new JsReturnStatement(expression))))
+                    bodyBlock: new JsBlockStatement(new JsReturnStatement(expression.WithAnnotation(ShouldBeObservableAnnotation.Instance)))
+                ));
+
+                if (expression.SatisfyResultCondition(n => n.HasAnnotation<ResultIsObservableAnnotation>()))
+                {
+                    arguments.Add(new JsLiteral(true));
+
+                    if (expression.SatisfyResultCondition(n => n.HasAnnotation<ResultIsObservableArrayAnnotation>()))
+                    {
+                        arguments.Add(new JsLiteral(true));
+                    }
+                }
+
+                return new JsIdentifierExpression("dotvvm").Member("evaluator").Member("wrapKnockoutExpression").Invoke(arguments)
                     .WithAnnotation(ResultIsObservableAnnotation.Instance)
                     .WithAnnotation(ShouldBeObservableAnnotation.Instance);
+            }
```

Comment:
I just think that it would be nicer to produce raw `ko.pureObservable` in case the result is not observable. In that case you'd have to remove the `ShouldBeObservableAnnotation`.

Reply:
Yes. I think it's a good idea.